lib/core/src/Cardano/Wallet/Api/Server.hs

```diff
@@ -2035,42 +2039,49 @@ mkRewardAccountBuilder ctx wid withdrawal = do
             (getRawKey $ deriveRewardAccount @k pwdP rootK, pwdP)
 
     withWorkerCtx ctx wid liftE liftE $ \wrk -> do
-        case withdrawal of
-           Nothing ->
-               pure (NoWithdrawal, selfRewardCredentials)
-
-           Just SelfWithdrawal -> do
-               (acct, _) <- liftHandler $ W.readRewardAccount @_ @s @k @n wrk wid
-               wdrl <- liftHandler $ W.queryRewardBalance @_ wrk acct
-               (, selfRewardCredentials) . WithdrawalSelf
-                   <$> liftIO (W.readNextWithdrawal @_ @k wrk wdrl)
-
-           Just (ExternalWithdrawal (ApiMnemonicT mw)) -> do
-               let (xprv, acct) = W.someRewardAccount @ShelleyKey mw
-               wdrl <- liftHandler (W.queryRewardBalance @_ wrk acct)
-                   >>= liftIO . W.readNextWithdrawal @_ @k wrk
-               when (wdrl == Coin 0) $ do
-                   liftHandler $ throwE ErrWithdrawalNotWorth
-               pure (WithdrawalExternal wdrl, const (xprv, mempty))
+        case (testEquality (typeRep @s) (typeRep @shelley), withdrawal) of
+            (Nothing, Just{}) ->
+                liftHandler $ throwE ErrReadRewardAccountNotAShelleyWallet
+
+            (_, Nothing) ->
+                pure (NoWithdrawal, selfRewardCredentials)
+
+            (Just Refl, Just SelfWithdrawal) -> do
+                (acct, path) <- liftHandler $ W.readRewardAccount @_ @s @k @n wrk wid
+                wdrl <- liftHandler $ W.queryRewardBalance @_ wrk acct
+                (, selfRewardCredentials) . WithdrawalSelf acct path
+                    <$> liftIO (W.readNextWithdrawal @_ @k wrk wdrl)
+
+            (Just Refl, Just (ExternalWithdrawal (ApiMnemonicT mw))) -> do
+                let (xprv, acct, path) = W.someRewardAccount @ShelleyKey mw
+                wdrl <- liftHandler (W.queryRewardBalance @_ wrk acct)
+                    >>= liftIO . W.readNextWithdrawal @_ @k wrk
+                when (wdrl == Coin 0) $ do
+                    liftHandler $ throwE ErrWithdrawalNotWorth
+                pure (WithdrawalExternal acct path wdrl, const (xprv, mempty))
```

Comment:
Almost no change here but the diff looks back because I re-indented things a bit. Main changes:

- Added `acct` and `path` to `WithdrawalSelf` and `WithdrawalExternal` 
- I also added an extra guard by checking the type of state this is running against. If it's not SeqState Shelley, we expect the withdrawal parameter to be necessarily `Nothing`, if not, it's an error (we can't withdraw on Icarus or Byron ... although our current API types would _allow_ that to go through).